### PR TITLE
Update klauspost/compress to fix data corruption at maximum compression

### LINKED ISSUE #4523: Corrupted backup with max compression on (Arch) Linux [restic 0.16.0]
Output of `restic version`
--------------------------

`restic 0.16.0 compiled with go1.20.7 on linux/amd64`

Edit: I now tested this also on MacOS, here is the output of restic version:

`restic 0.16.0 compiled with go1.20.6 on darwin/amd64`

How did you run restic exactly?
-------------------------------

I used the default environment variables. More info on the command i used
in the following sections. Note that I was trying to backup about 800 GB of
data, but i also tested this on the specific subfolder that was giving me problems.

The subfolder can be downloaded here:
https://drive.google.com/file/d/1O3XnVzEY6tNCVRo8_TmS1s5_JV0rGOGZ/view?usp=share_link

What backend/server/service did you use to store the repository?
----------------------------------------------------------------

I used a local repository, i tried storing the repository on different HDDs and NVMe to exclude
a possible hardware failure, but got the same result (with the same hashes) from restic check.

Expected behavior
-----------------

Restic check should not return any errors after backing up the folder.

Actual behavior
---------------

Restic check output:

```
using temporary cache in /tmp/restic-check-cache-4239199922
enter password for repository:
repository 952e5db1 opened (version 2, compression level auto)
created new cache in /tmp/restic-check-cache-4239199922
create exclusive lock for repository
load indexes
check all packs
check snapshots, trees and blobs
[0:00] 100.00%  1 / 1 snapshots
read all data
pack a7478fd99242cd86ddac79595494c07ff54165cf920554b49cab54c4946408fc contains 1 errors: [blob cf8f8df28d1de9f932ae567f10757c743bc9bf0020c636d27e98eae06b51df21: read blob <data/cf8f8df2> from a7478fd9: wrong data returned, hash is 2bc05a41d9c650277af6b02cd4bc69d6296a6ea532798e1062ec8fecb9dd3eb8]
pack fcd1c4bbdfe01bed5501125bc1a9704f08172dbaff92b632bb0cfe0243321aa2 contains 1 errors: [blob dc82d97be7683ecd41097ab02d7b15de81e8bbcd1c476c50b254b1f458090929: read blob <data/dc82d97b> from fcd1c4bb: wrong data returned, hash is 0b093d9ae6957ba8d3a1b09130623ba517e179347c3291af8d96fe954fd748f6]
[0:04] 100.00%  149 / 149 packs
Fatal: repository contains errors
```

Steps to reproduce the behavior
-------------------------------

- download files_to_backup.tar.gz from the provided google drive link
- unzip files_to_backup.tar.gz
- mkdir TEST_REPO
- restic -r TEST_REPO init
- restic -r TEST_REPO backup --compression max files_to_backup
- restic -r TEST_REPO check --read-data

Do you have any idea what may have caused this?
-----------------------------------------------

I don't exactly know, but without compression max it works on this folder, but
that may be just a coincidence.

Do you have an idea how to solve the issue?
-------------------------------------------

No.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

I love restic, and hope this will be fixed soon, thank you.

### PR DESCRIPTION
**Pull request by MichaelEischer**
What does this PR change? What problem does it solve?
-----------------------------------------------------
Fixes https://github.com/restic/restic/issues/4523 . The changelog entry assumes that https://github.com/restic/restic/pull/4530 has been merged.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added tests for all code changes.
- ~~[ ] I have added documentation for relevant changes (in the manual).~~
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

File: changelog/unreleased/issue-4523 [hunk 1]
```diff
@@ -0,0 +1,16 @@
+Bugfix: Update zstd library to fix possible data corruption at max. compression
+
+In restic 0.16.0, backups using maximum compression could result in data
+corruption. The "auto" compression level (used by default) is not affected!
```
Comment (rawtaz) [inline changelog/unreleased/issue-4523:4]: I suggest `.. could result in data corruption due to a bug in the library used for compressing data. Please note that the "auto" compression level (used by default) is not affected!`.

File: changelog/unreleased/issue-4523 [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+To check a repository for data corruption, run `restic check --read-data`. This
+will download and verify the whole repository. If the `check` command detects
+data corruption, follow the displayed steps. To simplify the repository repair
```
Comment (rawtaz) [inline changelog/unreleased/issue-4523:8]: s/displayed/suggested/

File: changelog/unreleased/issue-4523 [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+and minimize data loss, we've added an experimental `repair packs` command that
```
Comment (rawtaz) [inline changelog/unreleased/issue-4523:9]: s/we've added/we've also added/ ?

**Review by rawtaz — DISMISSED** [review]

File: changelog/unreleased/issue-4523 (hunk 1 above)
Comment (MichaelEischer) [reply changelog/unreleased/issue-4523]: I've applied the suggestions.

File: changelog/unreleased/issue-4523 [hunk 4]
```diff
@@ -0,0 +1,17 @@
+Bugfix: Update zstd library to fix possible data corruption at max. compression
+
+In restic 0.16.0, backups using maximum  could result in data corruption due to
```
Comment (rawtaz) [inline changelog/unreleased/issue-4523:3]: Missing "compression" in the middle of the sentence.

**Pull request merged**